engine/docker/docker.go

```diff
@@ -525,3 +537,33 @@ func (e *Docker) setProxyInDockerDaemon(ctx context.Context, pipelineConfig *spe
 		}
 	}
 }
+
+// SoftStop stops the container giving them a 30 seconds grace period. The signal sent by ContainerStop is SIGTERM.
+// After the grace period, the container is killed with SIGKILL.
+// After all the containers are stopped, they are removed only when the status is not "running" or "removing".
+func (e *Docker) SoftStop(ctx context.Context, name string) {
+	timeout := 30 * time.Second
+	if err := e.client.ContainerStop(ctx, name, &timeout); err != nil {
+		logrus.WithField("container", name).WithField("error", err).Warnln("failed to stop the container")
+	}
+
+	// Before removing the container we want to be sure that it's in a healthy state to be removed.
+
+	containerStatus, err := e.client.ContainerInspect(ctx, name)
+
+	if err != nil {
+		logrus.WithField("container", name).WithField("error", err).Warnln("failed to retrieve container stats")
+	}
+
+	for {
+		if err != nil && containerStatus.State.Status == "removing" || containerStatus.State.Status == "running" {
+			time.Sleep(1 * time.Second)
+			containerStatus, err = e.client.ContainerInspect(ctx, name)
```

Comment:
Can we move the `ContainerInspect` call only once inside the for loop and remove it from outside the for loop